[REVIEW] Migrating cuml comms to raft::comms

cjnolet opened the pull request:
This is based on top of initial work by @teju85 in #3. This is simply a copy (and reduction) of the current cuml comms to header-only form.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

cjnolet commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+    _requests_in_flight.insert(std::make_pair(*request, ucp_req));
+  }
+
+  void waitall(int count, request_t array_of_requests[]) const {
```
This synchronizes only on specific point-to-point messages (per the MPI interface) while `barrier()` synchronizes the entire cluster.

seunghwak replied:
waitall is wait on array_of_requests[] (you get a request for each non-blocking communication function call), this is different from communication barrier.

cjnolet replied:
Like the rest of the comms API, waitall is taken directly from the MPI API and is meant to synchronize for multiple point to point messages (which are always non-blocking in our API). We did this because we can implement blocking versions easily with non-blocking methods but the reverse is not true.

seunghwak commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+    get_request_id(request);
+    ucp_ep_h ep_ptr = (*_ucp_eps)[dest];
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
```
What  happens if malloc fails here?

cjnolet replied:
No call has been made to the UCX API so the sender would fail and any receiving ranks would timeout.

seunghwak replied:
If I understand this correctly, if malloc fails, it is better to throw an exception than continue execution.

I think
```
auto ucp_request = std::make_unique<ucp_request>();
```
will be safer.

seunghwak commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+      _requests_in_flight.erase(req_it);
+    }
+
+    while (requests.size() > 0) {
```
and UCX has no MPI_Waitall kinda function? Should we implement this by ourselves?

cjnolet replied:
UCX is a layer below MPI. This is at the UCP layer, which only knows endpoints and asynchronous message receiving/callbacks. It’s up to the user to implement blocking functions for multiple messages.

seunghwak replied:
I see, there is MPI libraries with UCX backend, and they may have exhaustively tuned their code for this, but IIRC, we have a reason we cannot use MPI, right? Little uneasy about implementing this by ourselves, but I guess we cannot avoid this to move forward :-)

seunghwak commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+      }
+
+      // Let other threads (including NCCL threads) use the CPU.
+      pthread_yield();
```
Better use standard C++ function (https://en.cppreference.com/w/cpp/thread/yield) than pthread?

cjnolet replied:
This was taken directly from the NCCL docs

seunghwak replied:
I see, but I think it is better to go for language standard than a third-party (not sure we can call pthread third-party but anyways) library.

seunghwak commented on cpp/include/raft/comms/ucp_helper.hpp:
```diff
@@ -0,0 +1,231 @@
+  /**
+   * @brief Frees any memory underlying the given ucp request object
+   */
+  void free_ucp_request(ucp_request *request) const {
```
This will lead to memory leak if users forget to call free_ucp_request or on exceptions. Can't we use std::unique_ptr with custom deleter doing what this function does instead?

cjnolet replied:
The comms api is meant to follow closely to MPI so that we can use CUDA-aware MPI underneath does automates gtests (which be in a follow-on PR). Users are not expected to know about, nor free UCX requests on their own. The requests are freed automatically when waitall is called. This is similar to MPI.

seunghwak replied:
I see, and is this the reason we're using malloc in

https://github.com/rapidsai/raft/pull/7/files/d4aa5c5cc58ba8754fa8d61bb95132b572c223cd#diff-0835dc79c2c9697632a30a73323b374fR232

?

I don't know the mechanism to automate gtests, so if there is a clear reason we cannot avoid using malloc, please ignore my comment about malloc above (comments explaining why we should stick to old style exception unsafe malloc will be helpful).

seunghwak commented on cpp/include/raft/cudart_utils.h:
```diff
@@ -112,14 +112,14 @@ class exception : public std::exception {
+  } while (0)
 
 /** helper method to get max usable shared mem per block parameter */
 inline int get_shared_memory_per_block() {
```
Isn't this redundant to https://github.com/rapidsai/raft/blob/branch-0.14/cpp/include/raft/handle.hpp#L157

cjnolet replied:
I have no idea. I just pulled this in because it was a dependency of the current comms.

seunghwak commented on cpp/include/raft/handle.hpp:
```diff
@@ -148,11 +149,17 @@ class handle_t {
-  //   std::shared_ptr<MLCommon::cumlCommunicator> communicator);
-  // const MLCommon::cumlCommunicator& getCommunicator() const;
-  // bool commsInitialized() const;
+  void set_comms(std::shared_ptr<comms::comms_t> communicator) {
```
Can a handle hold multiple (sub-)communicators?

cjnolet replied:
At the present time, I don’t relieve so, and I think we should discuss what that would like like in a follow-on PR, what do you think?

seunghwak replied:
Sure, no need to address everything in a single PR, just want to say that cuGraph needs sub-communicators.

cjnolet replied:
I'll open RAFT issues for the feature requests in this PR and someone from cugraph/cuml can knock them out as needed.

cjnolet commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,154 @@
+
+  virtual void reducescatter(const void* sendbuff, void* recvbuff,
+                             int recvcount, datatype_t datatype, op_t op,
+                             cudaStream_t stream) const = 0;
```
the assumption, at least on ML, is that these should always be CUDA Pointers and we should be doing as little as possible on host.

seunghwak replied:
Yes, if the data resides in GPU memory, we should avoid moving data back to host memory.

But say we are doing something like the following.
```
float error_sum = thrust::reduce(...);
comms.reduce(&error_sum, &error_sum, 1);
```
If our communicator does not support host memory, we need to transfer error_sum back to GPU memory, run comms.reduce(), then copy-back to CPU memory to figure out error is small enough or not.

More code, longer latency. I don't think this is desirable.

cjnolet replied:
It sounds like there are two different concerns here.  
*More code: * This sounds like a convenience feature and I'm not opposed to it. I'd prefer to open a Github issue and for someone to tackle it after the initial merge, though. What do you think?

*Longer latency: * Unfortunately, this is unavoidable as both UCX and NCCL are going to stage on host when a GPU-based transport is not available (eg. over network w/o infiniband/gpudirect). Wouldn't pinned host memory also reduce the latency here?

seunghwak replied:
>  I'd prefer to open a Github issue and for someone to tackle it after the initial merge, though. What do you think?

Sure, no problem.

>  Longer latency: Unfortunately, this is unavoidable as both UCX and NCCL are going to stage on host when a GPU-based transport is not available (eg. over network w/o infiniband/gpudirect). Wouldn't pinned host memory also reduce the latency here?

Oh, I am not talking about data residing in GPU memory and using host memory as a temporary buffer. I am talking about data actually residing in host memory (e.g. thrust reduce function returns reduced value in host memory). If we are communicating data residing in host memory, no need to involve GPU, and it will be nice if communicator supports this.
```
template <typename value_t>
void Environment::reduce(
    Communicator comm, Target target,
    value_t* p_src, value_t* p_dst, size_t value_count, int root, ReduceOp reduce_op,
    cudaStream_t stream) {
  MPI_Comm* p_mpi_comm{nullptr};
  ncclComm_t* p_nccl_comm{nullptr};
  std::tie(p_mpi_comm, p_nccl_comm) = get_comm_ptrs(comm);

  if (target == Target::HOST) {
    auto mpi_op = get_mpi_reduce_op(reduce_op);
    if (p_src == p_dst) {
      int comm_rank{};
      MPI_TRY(MPI_Comm_rank(*p_mpi_comm, &comm_rank));
      if (comm_rank == root) {
        MPI_TRY(
          MPI_Reduce(
            MPI_IN_PLACE, p_dst, value_count, get_mpi_type<value_t>(), mpi_op, root, *p_mpi_comm));
      }
      else {
        MPI_TRY(
          MPI_Reduce(
            p_src, p_dst, value_count, get_mpi_type<value_t>(), mpi_op, root, *p_mpi_comm));
      }
    }
    else {
      MPI_TRY(
        MPI_Reduce(p_src, p_dst, value_count, get_mpi_type<value_t>(), mpi_op, root, *p_mpi_comm));
    }
  }
  else if (target == Target::DEVICE) {
    auto nccl_op = get_nccl_reduce_op(reduce_op);
    NCCL_TRY(
      ncclReduce(
       p_src, p_dst, value_count, get_nccl_type<value_t>(), nccl_op, root,
       *p_nccl_comm, stream));
  }
  else {
    CUGRAPH_FAIL("Unsupported target");
  }
}
```
I have done something like the above (and we can move "Target target" as a template parameter to resolve this in compile time") to make communication library functions work for both host memory data and device memory.

No need to implement this in this PR, but we need to eventually add this support (and if we cannot use MPI for python/dask interaction, we need to implement this with UCX, but AFAIK, UCX figures out memory location from address, so we can largely reuse our UCX backend code for both host & device, correct?)

cjnolet replied:
That's correct- UCX supports this already but won't provide the collectives. We weren't able to get MPI working in Dask because of some of the assumptions it makes with regards to communicator lifecycles within a process. 

I suppose one option could be for us to create a feature request with the NCCL project. Not sure if we want to re-implement ring- and tree-style reductions w/ UCX, though I suppose that's another option.

seunghwak replied:
Reimplementing collectives using UCX sounds like an insane amount of work for us and largely repeating what MPI folks already did, better avoid unless there is really no other way.

It will be very nice if NCCL supports this :-)

cjnolet commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,154 @@
+
+  virtual status_t syncStream(cudaStream_t stream) const = 0;
+
+  virtual void isend(const void* buf, int size, int dest, int tag,
```
This follows the MPI API, which uses a 32-bit integer for this field.

seunghwak replied:
MPI uses int, but it is more like MPI's technical debt, and in theory, we'd better not follow their mistake.

In practice, if we rely on MPI to implement isend and if the size exceeds std::numerical_limit<int>::max(), we need to split a single isend call to multiple MPI non-blocking calls and we need to manage multiple request_t objects, and this will make our public API messier.

I suggest this. Use size_t size and if the backend is UCX or NCCL, this should just work. If the backend is MPI, if the size exceeds std::nemerical_limit<int>::max(), this call throws an exception; in this case, users will be responsible for splitting one large isend call to multiple isend calls.

cjnolet replied:
Just to be clear- I really don't mind this change at all, and it looks like both UCX and NCCL support `size_t` as well.

cjnolet commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,154 @@
+  }
+
+ private:
+  std::unique_ptr<comms_iface> _impl;
```
I ran clang-format and am unaware that we are supposed to be using clang tidy. I really just haven’t been keeping up on the discussions. I can run it if that’s what’s been agreed upon.

seunghwak replied:
@teju85 AFAIK, we're expected to run clang-tidy for raft code, could you confirm this?

cjnolet commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,154 @@
+namespace comms {
+
+typedef unsigned int request_t;
+enum datatype_t { CHAR, UINT8, INT, UINT, INT64, UINT64, FLOAT, DOUBLE };
```
@jirikraus designed the original communicator in cuml and I understand he’s spent a lot of time in the cuda-aware MPI community at Nvidia. I’m tagging him for his thoughts here.

cjnolet commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,154 @@
+                             cudaStream_t stream) const = 0;
+};
+
+class comms_t : public comms_iface {
```
We have an MPI+NCCL  implementation of the comms that allows us to run and test algorithms without the need for bootstrapping UCX in Dask. This allows the handle to hold a comms object without having to be concerned with the implementation. If there’s a better way of doing this in a header-only library, I’m certainly open to alternatives, but I would like it if we kept the implementation of the comms separate from the interface.

seunghwak replied:
I am not sure I am following this. Absolutely agree with "we kept the implementation of the comms separate from the interface." What I don't understand is this class has a comms_iface object as a member variable and calls comms_iface functions through this member variable and does nothing else besides this. Anything breaks if we remove "public comms_iface" from the code above?

I understand that std_comms should inherit comms_iface as std_comms implements virtual methods in comms_iface, but I don't get why comms_t here needs to inherit comms_iface.

cjnolet replied:
Indeed. I was responding to you on my iphone last night and overlooked this detail. This was an oversight- since there's a conversion constructor from `comms_iface`, this should no longer have to extend `comms_iface`.

cjnolet commented on cpp/include/raft/comms/comms_helper.hpp:
```diff
@@ -0,0 +1,146 @@
+namespace raft {
+namespace comms {
+
+void build_comms_nccl_only(handle_t *handle, ncclComm_t comm, int size,
```
The reason for these factory-like methods is to abstract the injection of the conms on the handle from Python, rather than having to expose these methods on the handle. We could do that, but it seems cleaner to have cython call these functions. Otherwise, I do agree, but I wouldn’t expect someone in the c++ layer to ever be calling  these functions directly.

seunghwak replied:
Got it, makes sense if this is mainly for cython interface.

cjnolet replied:
I could rename the file to something like `cython_helper.hpp`, but I'd prefer that we not limit its potential use. What do you think about a comment in the code that specifies these functions help enable the cython layer?

seunghwak replied:
Yes, that comment will be very helpful. Comments are generally very helpful if we make some non-obvious design choices/work-arounds due to practical restrictions.

cjnolet commented on cpp/include/raft/comms/comms_helper.hpp:
```diff
@@ -0,0 +1,146 @@
+  handle->set_comms(communicator);
+}
+
+bool test_collective_allreduce(const handle_t &handle) {
```
Similar to the “build_*” functions, these are meant to be called by Cython. I lumped these functions all in this file to minimize the number of files we need to maintain. I can certainly create different file if that’s preferred, but these functions are not gtests, so they would still need to be available in this API.

seunghwak replied:
I see, so python actually runs this on initialization to check this the communication interface is working, then, this makes much more sense.

cjnolet replied:
Python doesn't run it on initialization, but it is exposed to Python as a function that can be called in the shell. It's also exposed to pytests for our automated testing. I don't mind moving it, I'm just trying to minimize the number of files we need to keep around. Perhaps we could find a better name for `comms_helper.hpp`?

seunghwak replied:
OK, let's leave as is for now, this sounds like a minor detail and we can deal with this later if necessary.

cjnolet commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+    initialize();
+  };
+
+  virtual ~std_comms() {
```
Originally this comms was a separate library in cuml, which didn’t have the device buffer available as a dependency (it would have created a circular dependency) so we just used cudamalloc. It’s only allocating a single byte for use in `barrier` but I do think device buffer would be cleaner.

cjnolet commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+  int getRank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> commSplit(int color, int key) const {
+    // Not supported by NCCL
```
Interesting. Sylvain always told me to avoid using more than one communicator in the same process (Eg same gpu) because it can lead to deadlocks when the streams are not able to schedule warps concurrently. Perhaps this changed in more recent versions? We might want to investigate this a little further to be sure.

seunghwak replied:
It may be problematic if we concurrently run multiple communicators (ideally this should work as well and this works pretty well for MPI but this may not be a case for NCCL yet), but if we communicate using different communicators in different phases, I haven't encountered any problem.

cjnolet replied:
Since you brought it up, I also want to point out that my other concern over this feature was related to high initialization latencies I was experiencing when creating new NCCL communicators. Contrary to the expected behavior, these latencies were not being reduced at all in subsequent initializations.

@divyegala noticed in some of his recent benchmarking that this appears to have been fixed.

seunghwak replied:
Yeah... I've seen that issue as well. But IIRC, this overhead gets worse with process count, so creating sub-communicators is  generally cheaper than creating a full communicator.

Nonetheless no need to pay unnecessary overhead if sub-communicators are unnecessary, and we may better not initialize sub-communicators by default but better provide an option.

cjnolet commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+
+#include <raft/cudart_utils.h>
+
+#define NCCL_CHECK(call)                                                       \
```
Created Issue #11

cjnolet commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+
+  void irecv(void *buf, int size, int source, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
```
I like this. Where do you think this should live?

seunghwak replied:
These are copy-and-paste-and-modify of https://github.com/rapidsai/cudf/blob/branch-0.14/cpp/include/cudf/utilities/error.hpp, and I think eventually, all these should go to rapids-core.

At this moment, just follow cudf, and place under include/raft/error.hpp or include/raft/utilities/error.hpp?

seunghwak replied:
And another question is whether we should place CUML_EXPECTS, CUML_FAIL, CUGRAPH_EXPECTS, CUGRAPH_FAIL in raft or cuml & cudf. If we place those here, it will be helpful in enforcing consistent error handling mechanism and minimizing code duplicates.

afender replied:
Yes, I was suggesting having a `RAFT_EXPECT` and other error macros here (like `NCCL_TRY`) so that we can use this consistently across RAFT, cuML and cuGraph repos. 
I don't think rapids-core is happening anytime soon so better move forward with this now through RAFT.

teju85 replied:
I have a concern with this. The current *_EXPECT does not seem to allow formatting of string to provide more meaningful error messages. For eg: We should allow devs to do something like this:
` RAFT_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch rhs=%s lhs=%s", to_string(lhs->dtype), to_string(rhs->dtype));`

I know that core-C++ folks might cringe at C-style modifiers, I'm ok if we decide to make it more C++-ish.

seunghwak replied:
Can't we do "Column type mismatch rhs=" + to_string(lhs->dtype) + " lhs=" + to_string(rhs->dtype)? Need to check this compiles but there must be a way to do this in a C++ way then relying on the old C style mechanism (this sometimes leads to a weird bug if type specifier and actual type mismatch). I need to check RAFT_EXPECTS internals but we may be able to make "Column type mismatch rhs=" << lhs->dtype << " lhs=" << lhs->dtype work as well to avoid to_string calls (if not already works).

afender replied:
We may be able to reconcile, but consistency with cuDF is nice too. Let's continue the discussion in #11.

cjnolet commented:
@seunghwak @afender I've implemented much of your requested changes, including adding the templates back into the `comms_t` functions. As discussed already, I propose we hold the unified exceptions & generic nccl_type/nccl_op wrappers for a follow-on PR. I have also added more docs and pytests to verify the collectives end-to-end. Please let me know if you are okay merging this in.

teju85 commented:
@cjnolet Sorry, I haven't been able to look at this PR till now. Give me another day to take a look at this.

cjnolet commented:
@teju85, will do! I think I owe you a review as well. Let me get started on that if someone else hasn't been able to do it already.

seunghwak commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,327 @@
+ public:
+  virtual ~comms_iface();
+
+  virtual int getSize() const = 0;
```
https://github.com/rapidsai/raft/blob/branch-0.14/cpp/.clang-tidy#L143

Doesn't the above rule apply here?

cjnolet replied:
For some reason, I am not able to execute @teju85's `run-clang-tidy` script and I don't see any instructions in the codebase. I was able to run it manually, but I'm not completely sure I did it properly.

teju85 replied:
I think it is better to enable clang-tidy in CI, now that we have it ready in cuML too?

afender commented:
> Please let me know if you are okay merging this in.

Thanks, Corey. It looks great to me :+1: . Lots of good discussions on this PR! 
Moving forward, I also think it is a good time to merge it. This will allow incrementally building on top while proposing upgrades as needed.

cjnolet commented:
It looks like there's a bug in GCC5 in CI.  @dantegd has tracked down the bug, but it sounds like it might be worth investigating a workaround or just having `get_type()` evaluate at runtime until we can find one. Do you guys have any thoughts?

seunghwak commented:
> It looks like there's a bug in GCC5 in CI. @dantegd has tracked down the bug, but it sounds like it might be worth investigating a workaround or just having `get_type()` evaluate at runtime until we can find one. Do you guys have any thoughts?

What was the error message like? Was it a compile time error?

https://github.com/rapidsai/raft/pull/7/files#diff-6026184287025decc1dddbc2c7a2386fR96

May need to add constexpr here as well?

Or we can implement the same without template specialization if this work arounds the problem.
```
template <typename value_t>
constexpr data_type_t get_type() {
  if (std::is_integral<value_t>::value) {
    if (sizeof(value_t) == 1) {
      if (std::is_signed<value_t>::value) {
        return datatype_t::CHAR;
      }
      esle {
        return datatype_t::UCHAR;
      }
    }
    else if (sizeof(value_t) == 2) {
    }
    ...
  }
  else if(...) {
    ...
  }
  ...
}
```

dantegd commented:
@seunghwak the error is in the cuda10.1 CI (that uses G++ 5):

```bash
In file included from /var/lib/jenkins/workspace/rapidsai/gpuci/raft/prb/raft-gpu-build_3/cpp/include/raft/handle.hpp:37:0,
                 from /var/lib/jenkins/workspace/rapidsai/gpuci/raft/prb/raft-gpu-build_3/cpp/test/handle.cpp:20:
/var/lib/jenkins/workspace/rapidsai/gpuci/raft/prb/raft-gpu-build_3/cpp/include/raft/comms/comms.hpp: In member function 'raft::comms::datatype_t raft::comms::comms_t::get_type() const [with value_t = char]':
/var/lib/jenkins/workspace/rapidsai/gpuci/raft/prb/raft-gpu-build_3/cpp/include/raft/comms/comms.hpp:288:22: error: enclosing class of constexpr non-static member function 'raft::comms::datatype_t raft::comms::comms_t::get_type() const [with value_t = char]' is not a literal type
 constexpr datatype_t comms_t::get_type<char>() const {
                      ^
/var/lib/jenkins/workspace/rapidsai/gpuci/raft/prb/raft-gpu-build_3/cpp/include/raft/comms/comms.hpp:89:7: note: 'raft::comms::comms_t' is not literal because:
 class comms_t {
       ^
/var/lib/jenkins/workspace/rapidsai/gpuci/raft/prb/raft-gpu-build_3/cpp/include/raft/comms/comms.hpp:89:7: note:   'raft::comms::comms_t' has a non-trivial destructor
```

which doesn't happen in the G++ 7 based jobs (all of this reproducible outside docker with the correct gcc version). After tracking the error it seems to be this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66297 which was solved in 7.2. Note: It only affects that part of the code/pattern, so having c++14 is fine except for that.

dantegd commented:
It was a compile time error at the setup.py/compiling cython objects moment for further information.

seunghwak commented:
> It was a compile time error at the setup.py/compiling cython objects moment for further information.

Is there a reason get_datatype() needs to be a member function of comms_t?

"'raft::comms::comms_t' has a non-trivial destructor" may go out if we move the function out?

teju85 commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+  }
+}
+
+ncclDataType_t getNCCLDatatype(const datatype_t datatype) {
```
Not too sure about this @afender . We are passing `datatype` at runtime (coming from the comms interface itself) and thus not sure how this would be helpful.

teju85 replied:
BTW, for the opposite direction, however, there's already something similar (`get_type` method)

afender replied:
That's fine.

Taking a step back, we should look at the motivation for the `datatype_t. We had something similar in cuGraph a few releases back for the C-API but that went away as part of C++ upgrade.

This didn't really bring limitations while allowing to better leverage native C++ types, structures, templates, and std:: functions. As @seunghwak pointed out, the emergence of device_vector to replace void buffers will further reduce the need for carrying custom enum type as well.

teju85 commented on cpp/include/raft/comms/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+#
```
a cmake file inside include dir?

teju85 commented on cpp/include/raft/comms/comms.hpp:
```diff
@@ -0,0 +1,331 @@
+ * The resulting status of distributed stream synchronization
+ */
+enum class status_t {
+  commStatusSuccess,  // Synchronization successful
+  commStatusError,    // An error occured querying sync status
+  commStatusAbort     // A failure occurred in sync, queued operations aborted
```
this is odd. clang-tidy should have caught this. Not sure what's happening here!

teju85 commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,406 @@
+namespace raft {
+namespace comms {
+
+size_t getDatatypeSize(const datatype_t datatype) {
```
agreed!

teju85 commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,417 @@
+  }
+}
+
+ncclDataType_t get_nccl_datatype(const datatype_t datatype) {
```
same as above `constexpr` should be useful here too

teju85 commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,417 @@
+  }
+}
+
+ncclRedOp_t get_nccl_op(const op_t op) {
```
`constexpr`

teju85 commented on cpp/include/raft/comms/nccl_helper.hpp:
```diff
@@ -0,0 +1,34 @@
+
+namespace raft {
+namespace comms {
+inline void ncclUniqueIdFromChar(ncclUniqueId *id, char *uniqueId, int size) {
```
use `lower_case` format for function name

teju85 commented on cpp/include/raft/comms/std_comms.hpp:
```diff
@@ -0,0 +1,417 @@
+  }
+
+  void initialize() {
+    CUDA_CHECK(cudaStreamCreate(&stream_));
```
bit confused here. Why do we need a separate `stream_` member when all other functions are accepting `stream` parameter?

Iroy30 commented on python/raft/common/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# limitations under the License.
+#
+
+from raft.common.cuda import Stream
```
I believe these imports should be relative

Iroy30 commented on python/raft/dask/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# limitations under the License.
+#
+
+from raft.dask.common.comms import Comms
```
Relative imports

Iroy30 commented on python/raft/dask/common/__init__.py:
```diff
@@ -0,0 +1,26 @@
+# limitations under the License.
+#
+
+from raft.dask.common.comms import Comms
```
Relative imports

Iroy30 commented on python/raft/dask/common/comms.py:
```diff
@@ -0,0 +1,412 @@
+# limitations under the License.
+#
+
+from raft.dask.common.nccl import nccl
```
Relative imports

Iroy30 replied:
Relative imports were missed.

cjnolet replied:
Thanks  looking again, Ishika. I've fixed this and gone through to verify the only remaining absolute imports are in the pytests.

cjnolet commented:
@teju85 @Iroy30 I have implemented your requested changes. Mind giving this PR a re-review?

cjnolet commented:
@seunghwak @afender, making the nccl type and data type size functions static fixes the GCC issue for now. I can create a new issue for us to knock out in the future so that we can get this merged soon. What do you think?

seunghwak commented:
> @seunghwak @afender, making the nccl type and data type size functions static fixes the GCC issue for now. I can create a new issue for us to knock out in the future so that we can get this merged soon. What do you think?

Sounds good for me, thanks for the effort!

teju85 commented on cpp/CMakeLists.txt:
```diff
@@ -67,6 +67,8 @@ find_package(CUDA 10.0 REQUIRED)
 
 set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
 
+message("HELLO!")
```
this is probably a remnant of a debug session?

teju85 commented on cpp/cmake/comms.cmake:
```diff
@@ -0,0 +1,37 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+project(comms LANGUAGES CXX CUDA)
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
```
why do we need this module-path set command here, in a module?

cjnolet commented:
@teju85 Thank you for the re-review. I've addressed the latest comments. Is this ready to be merged?

cjnolet commented:
@dantegd, I understood you and Ishika both needed the relative Python imports. Do you have any further review feedback?

afender commented:
Synced up with Dante on Slack :+1: Ishika is going to add a PR on top of this on in the next few hours/days so we can keep adding enhancement there.

Merging :rocket:
